pkg/controllers/subnetset/subnetset_controller.go

```diff
@@ -0,0 +1,236 @@
+	newConditions := []v1alpha1.Condition{
+		{
+			Type:    v1alpha1.Ready,
+			Status:  v1.ConditionTrue,
+			Message: "NSX SubnetSet has been successfully created/updated",
+			Reason:  "SubnetsReady",
+		},
+	}
+	r.updateSubnetSetStatusConditions(ctx, subnetset, newConditions)
+}
+
+func (r *SubnetSetReconciler) setSubnetSetReadyStatusFalse(ctx *context.Context, subnetset *v1alpha1.SubnetSet) {
+	newConditions := []v1alpha1.Condition{
+		{
+			Type:    v1alpha1.Ready,
+			Status:  v1.ConditionFalse,
+			Message: "NSX SubnetSet could not be created/updated",
+			Reason:  "SubnetNotReady",
+		},
+	}
+	r.updateSubnetSetStatusConditions(ctx, subnetset, newConditions)
+}
+
+func (r *SubnetSetReconciler) updateSubnetSetStatusConditions(ctx *context.Context, subnetset *v1alpha1.SubnetSet, newConditions []v1alpha1.Condition) {
+	conditionsUpdated := false
+	for i := range newConditions {
+		if r.mergeSubnetSetStatusCondition(ctx, subnetset, &newConditions[i]) {
+			conditionsUpdated = true
+		}
+	}
+	if conditionsUpdated {
+		r.Client.Status().Update(*ctx, subnetset)
+		log.V(1).Info("updated Subnet", "Name", subnetset.Name, "Namespace", subnetset.Namespace,
+			"New Conditions", newConditions)
+	}
+}
+
+func (r *SubnetSetReconciler) mergeSubnetSetStatusCondition(ctx *context.Context, subnetset *v1alpha1.SubnetSet, newCondition *v1alpha1.Condition) bool {
+	matchedCondition := getExistingConditionOfType(newCondition.Type, subnetset.Status.Conditions)
+
+	if reflect.DeepEqual(matchedCondition, newCondition) {
+		log.V(2).Info("conditions already match", "New Condition", newCondition, "Existing Condition", matchedCondition)
+		return false
+	}
+
+	if matchedCondition != nil {
+		matchedCondition.Reason = newCondition.Reason
+		matchedCondition.Message = newCondition.Message
+		matchedCondition.Status = newCondition.Status
+	} else {
+		subnetset.Status.Conditions = append(subnetset.Status.Conditions, *newCondition)
+	}
+	return true
+}
+
+func getExistingConditionOfType(conditionType v1alpha1.ConditionType, existingConditions []v1alpha1.Condition) *v1alpha1.Condition {
+	for i := range existingConditions {
+		if existingConditions[i].Type == conditionType {
+			return &existingConditions[i]
+		}
+	}
+	return nil
+}
+
+func (r *SubnetSetReconciler) setupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1alpha1.SubnetSet{}).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: runtime.NumCPU(),
+		}).
+		Watches(&source.Kind{Type: &v1alpha1.VPC{}},
+			&VPCHandler{Client: mgr.GetClient()},
+			builder.WithPredicates(VPCPredicate)).
+		Watches(
+			&source.Kind{Type: &v1alpha1.SubnetPort{}},
+			&SubnetPortHandler{Reconciler: r},
+			builder.WithPredicates(SubnetPortPredicate)).
+		Complete(r)
+}
+
+// GarbageCollector collect Subnet which there is no port attached on it.
+// cancel is used to break the loop during UT
+func (r *SubnetSetReconciler) GarbageCollector(cancel chan bool, timeout time.Duration) {
```

Comment:
seems GC can delete subnets which do not have ports on it, can it also handle there is no SubnetSet CR in K8s, but in nsx  subnet store, still has the NSX subnets for the SubnetSet CR?

Reply:
> seems GC can delete subnets which do not have ports on it, can it also handle there is no SubnetSet CR in K8s, but in nsx subnet store, still has the NSX subnets for the SubnetSet CR?

No, it can't handle this case currently. We can also mark a TODO here.